vividus-plugin-web-app/src/main/java/org/vividus/proxy/ThreadedProxy.java

```diff
@@ -16,73 +16,146 @@
+import org.slf4j.LoggerFactory;
+import org.vividus.model.IntegerRange;
+
 public class ThreadedProxy implements IProxy
 {
-    @Inject private IProxyServerFactory proxyServerFactory;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThreadedProxy.class);
+
+    private final InetAddress proxyHost;
+    private final Queue<Integer> proxyPorts;
+    private final boolean useEphemeralPort;
+    private final ThreadLocal<IProxy> proxy;
+
+    public ThreadedProxy(String proxyHost, IntegerRange proxyPorts, IProxyFactory proxyFactory)
+            throws UnknownHostException
+    {
+        List<Integer> proxyPortsRange = proxyPorts.getRange();
+        if (proxyPortsRange.contains(0))
+        {
+            Validate.isTrue(proxyPortsRange.size() == 1,
+                    "Port 0 (ephemeral port selection) can not be used with custom ports");
+            useEphemeralPort = true;
+        }
+        else
+        {
+            useEphemeralPort = false;
+            proxyPortsRange.forEach(ThreadedProxy::validatePort);
+        }
+        this.proxyPorts = new LinkedList<>(proxyPortsRange);
+        this.proxy = ThreadLocal.withInitial(proxyFactory::createProxy);
+        this.proxyHost = InetAddress.getByName(proxyHost);
+    }
 
-    private final ThreadLocal<IProxy> proxy = ThreadLocal.withInitial(() ->
+    @SuppressWarnings("MagicNumber")
+    private static void validatePort(int port)
     {
-        Proxy proxy = new Proxy();
-        proxy.setProxyServerFactory(proxyServerFactory);
-        return proxy;
-    });
+        Validate.isTrue(port > 0 && port <= 65535, "Expected ports range is 1-65535 but got: %d", port);
+    }
 
     @Override
     public void start()
     {
-        proxy.get().start();
+        if (useEphemeralPort)
+        {
+            proxy().start();
+            return;
+        }
+        synchronized (proxyPorts)
+        {
+            Validate.isTrue(!proxyPorts.isEmpty(), "There are no available ports in the ports pool");
+            int port = proxyPorts.poll();
+            LOGGER.atInfo().log("Allocate {} port from the proxy ports pool", port);
+            logAvailablePorts();
+            proxy().start(port, proxyHost);
+        }
+    }
+
+    @Override
+    public void start(int port, InetAddress address)
+    {
+        proxy().start(port, address);
     }
 
     @Override
     public void startRecording()
     {
-        proxy.get().startRecording();
+        proxy().startRecording();
     }
 
     @Override
     public void stopRecording()
     {
-        proxy.get().stopRecording();
+        proxy().stopRecording();
     }
 
     @Override
     public void stop()
     {
-        proxy.get().stop();
+        if (!useEphemeralPort)
+        {
+            synchronized (proxyPorts)
+            {
+                int port = proxy().getProxyServer().getPort();
+                proxyPorts.add(port);
+                LOGGER.atInfo().log("Return {} port back to the proxy ports pool", port);
```

Comment:
`LOGGER.atInfo().addArgument(port).log("Return {} port back to the proxy ports pool");`